src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -21706,6 +21706,484 @@ void M(string? s)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1, p1).ToString();
+        c.M(p1, p2).ToString(); // 1
+        c.M(p2, p1).ToString();
+        c.M(p2, p2).ToString(); // 2
+    }
+}";
+            var expected = new[]
+            {
+                // (8,9): warning CS8602: Dereference of a possibly null reference.
+                //         c.M(p1, p2).ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.M(p1, p2)").WithLocation(8, 9),
+                // (10,9): warning CS8602: Dereference of a possibly null reference.
+                //         c.M(p2, p2).ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.M(p2, p2)").WithLocation(10, 9)
+            };
+
+            var lib = CreateNullableCompilation(new[] { NotNullIfNotNulltAttributeDefinition, lib_cs });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+            CompileAndVerify(comp2);
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_Return_MultipleParameters_SameName()
+        {
+            var source =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""p"")] public string? M(string? p, string? p) => throw null!;
+
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1, p1).ToString();
+        c.M(p1, p2).ToString();
+        c.M(p2, p1).ToString();
+        c.M(p2, p2).ToString(); // 1
+    }
+}";
+            var comp2 = CreateNullableCompilation(new[] { source, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(
+                // (4,73): error CS0100: The parameter name 'p' is a duplicate
+                //     [return: NotNullIfNotNull("p")] public string? M(string? p, string? p) => throw null!;
+                Diagnostic(ErrorCode.ERR_DuplicateParamName, "p").WithArguments("p").WithLocation(4, 73),
+                // (12,9): warning CS8602: Dereference of a possibly null reference.
+                //         c.M(p2, p2).ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.M(p2, p2)").WithLocation(12, 9)
+                );
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_Return_NonExistentName()
+        {
+            var source =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""missing"")] public string? M(string? p) => throw null!;
+
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1).ToString(); // 1
+        c.M(p2).ToString(); // 2
+    }
+}";
+            // Warn on misused attribute? https://github.com/dotnet/roslyn/issues/36073
```

Comment:
Realistically, we need to do this now or it'll have to go in warning wave.

Reply:
Correct. We've discussed this with Mads and the set of warnings for misused attributes won't make it into C# 8. There are subtleties...

---
In reply to: [303561371](https://github.com/dotnet/roslyn/pull/37201#discussion_r303561371) [](ancestors = 303561371)